### PR TITLE
[Distillation] base learn-to-init llama attention for distillation

### PR DESCRIPTION
**Pull request by vlad-karp**
# Description

This PR introduces the base implementation for Learn-to-Init (LTI) attention for distillation (llama only be can easily be generalized). 

Relevant details and context:
- **Problem being solved:** Setting up the foundational LTI components required for effective attention distillation of LLaMA models.
- **Implementation details:** 
  - Adds the core LTI logic in a new `learn_to_init_layer.py` module.
  - Updates the distillation pipeline (`distillation_utils.py` and `train_distill.py`) and decoders to support the new attention layer.
  - 2 LTI implementation for GQA - using bi-linear and global linear map options.
  - Configures the system to get LTI student init-time teacher shapes directly from the configuration.

# Tests
- Added a new unit test for LearnToInitDense as well as teacher to student weight injection logic.
src/maxtext/tests/post_training/unit/learn-to-init_test.py

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/models/learn_to_init_layer.py [hunk 1]
```diff
@@ -0,0 +1,467 @@
...
+      if isinstance(child, linears.DenseGeneral):
+        orig_proj_shape = child.kernel.shape
+        assert len(orig_proj_shape) == 3
+        if name in ("query", "key", "value"):
+          teacher_heads_num = teacher_config.base_num_query_heads if name == "query" else teacher_config.base_num_kv_heads
+          teacher_shape = (orig_proj_shape[0], teacher_heads_num, teacher_config.head_dim)
+        elif name == "out":
+          teacher_shape = (teacher_config.base_num_query_heads, teacher_config.head_dim, orig_proj_shape[2])
+        else:
+          # Should not happen with the current target_names
```
Comment (entrpn) [inline src/maxtext/models/learn_to_init_layer.py]: better to log a warning here just in case.

File: src/maxtext/models/learn_to_init_layer.py [hunk 2]
```diff
@@ -0,0 +1,467 @@
...
+    self.kernel_init = kernel_init
+    self.kernel_axes = kernel_axes
+    self.quant = quant
+    self.use_bias = use_bias
+    self.shard_mode = shard_mode
+    self.matmul_precision = matmul_precision
+    self.parameter_memory_host_offload = parameter_memory_host_offload
+    self.use_general_linear_map = use_general_linear_map
+
+    self.C = nnx.Param(C, sharding=())
```
Comment (entrpn) [inline src/maxtext/models/learn_to_init_layer.py]: wouldn't setting up an empty sharding here blow up the memory as C would be replicated. Same applies to the nnx.Params below.

File: src/maxtext/models/llama2.py [hunk 3]
```diff
@@ -212,6 +212,13 @@ def __call__(
       return layer_output, kv_cache
 
 
+LlamaLTIDecoderLayer = functools.partial(LearnToInitDecoderLayer, base_layer_cls=LlamaDecoderLayer)
```
Comment (entrpn) [inline src/maxtext/models/llama2.py:215]: From types there is a new LLAMA2LIT. Is it possible to instead have a flag to enable learn to init and code a mapping for LTI such that its generic for any model?

Comment (vlad-karp) [reply src/maxtext/models/llama2.py:215]: was thinking about that - it feels lake a larger task. To make it more generic, we need to account for different types of attention blocks, for example.
The current implementation won't work, for example, for qkv fused attention mode. Not for MLA. 
I suggest we first make sure this approach is useful for qwen moe, then we can generalize it + inlclude MoE inversion LTI, etc.

File: src/maxtext/trainers/post_train/distillation/distillation_utils.py [hunk 4]
```diff
@@ -180,6 +180,62 @@ def __next__(self) -> MaxTextTrainingInput:
...
+    dest_node = student_graph.get(dest_path, None)
+    assert source_node is not None, f"Could not find source_node model parameter at path: {source_path}"
+    assert dest_node is not None, f"Could not find dest_node model parameter at path: {dest_path}"
+
+    assert source_node.value.shape == dest_node.value.shape, (
+        f"Shape mismatch for A sharing between {source_path} and {dest_path}: "
+        f"{source_node.value.shape} vs {dest_node.value.shape}"
+    )
+    max_logging.info(f"Sharing parameter {source_path} with {dest_path}")
+    student_graph[dest_path] = source_node
```
Comment (JamesDeng42) [inline src/maxtext/trainers/post_train/distillation/distillation_utils.py:223]: this is a bug, i think, we should do "dest_node.value = source_node.value".

Comment (vlad-karp) [reply src/maxtext/trainers/post_train/distillation/distillation_utils.py:223]: no, this is intentional - we share parameters INSIDE the same student model, so we need to use the same nnx.Parameter instance, not the (immutable) jax arrays

**Review by entrpn — APPROVED** [review]

File: src/maxtext/trainers/post_train/distillation/distillation_utils.py [hunk 5]
```diff
@@ -180,6 +180,62 @@ def __next__(self) -> MaxTextTrainingInput:
     )
 
 
+def prepare_student_weights(
```
Comment (gagika) [inline src/maxtext/trainers/post_train/distillation/distillation_utils.py]: Can we move it to a new file, lti_utils.py or learn_to_init_utils.py? As learn to init code will increase, let's try to separate them..

File: src/maxtext/layers/learn_to_init_layer.py [hunk 6]
```diff
@@ -0,0 +1,452 @@
+# Copyright 2023–2026 Google LLC
```
Comment (gagika) [inline src/maxtext/layers/learn_to_init_layer.py:1]: can we move from the models directory?

Maybe under src/maxtext/layers/

File: src/maxtext/trainers/post_train/distillation/distillation_utils.py [hunk 7]
```diff
@@ -600,9 +600,11 @@ def __init__(
       self,
       raw_iterator: Any | None,
       root_directory: str | None = None,
+      student_config: Any | None = None,
```
Comment (gagika) [inline src/maxtext/trainers/post_train/distillation/distillation_utils.py]: nit: can it be None?
perhaps change default and data type if it can't.

File: src/maxtext/layers/learn_to_init_layer.py [hunk 8]
```diff
@@ -0,0 +1,439 @@
...
+  """
+  Applies the finalized learn-to-init weights to the student model and cleans up the NNX graph.
+
+  This function iterates over the `LearnToInitDense` layers in the trained student model,
+  calculates their final, static effective kernels using `calculate_attn_weight`, and
+  replaces the dynamically-computed LTI modules with standard kernel representations.
+  It effectively collapses the learn-to-init parameterization back into a standard
+  decoder architecture, modifying the `student_model` in-place.
+
+  NOTE: works for ToNXX decoder model and layer-scan mode only
```
Comment (gagika) [inline src/maxtext/layers/learn_to_init_layer.py:376]: could you throw an exception if it's not layer-scan mode?

File: src/maxtext/common/common_types.py [hunk 9]
```diff
@@ -108,6 +108,8 @@ class DecoderBlockType(enum.Enum):
   LLAMA4 = "llama4"
   OLMO3 = "olmo3"
 
+  LLAMA2LTI = "llama2-learn-to-init"
```
Comment (gagika) [inline src/maxtext/common/common_types.py]: for naming convention could you use "_" instead of "-" e.g. LLAMA2_LTI = "llama2_lti"

File: tests/post_training/unit/learn_to_init_test.py
Comment (gagika) [inline tests/post_training/unit/learn_to_init_test.py:1]: could you name the test with "_" insead of "-"? e..g. learn_to_init_test.py

File: src/maxtext/trainers/post_train/distillation/train_distill.py [hunk 10]
```diff
@@ -366,6 +367,7 @@ def _post_process_train_step(self, aux: dict[str, tuple[jax.Array, jax.Array]])
         self._buffered_train_metrics.additional_metrics[name] = ([], distillation_utils.weighted_mean)
 
       self._buffered_train_metrics.additional_metrics[name][0].append(value)
+    max_logging.log(f"Distillation metrics: {aux}")
```
Comment (gagika) [inline src/maxtext/trainers/post_train/distillation/train_distill.py:379]: is it logged at every step or once? as it's inside _post_process_train_step

**Review by gagika — APPROVED** [review]
